core/state_processor_rip7560.go

```diff
@@ -346,12 +344,32 @@ func ApplyRip7560ExecutionPhase(config *params.ChainConfig, vpr *ValidationPhase
 	accountExecutionMsg := prepareAccountExecutionMessage(vpr.Tx, evm.ChainConfig())
 	executionResult, err := ApplyMessage(evm, accountExecutionMsg, gp)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	var paymasterPostOpResult *ExecutionResult
 	if len(vpr.PaymasterContext) != 0 {
 		paymasterPostOpResult, err = applyPaymasterPostOpFrame(vpr, executionResult, evm, gp, statedb, header)
 	}
+	return executionResult, paymasterPostOpResult, err
+}
+
+func ApplyRip7560ExecutionPhase(config *params.ChainConfig, vpr *ValidationPhaseResult, bc ChainContext, author *common.Address, gp *GasPool, statedb *state.StateDB, header *types.Header, cfg vm.Config) (*types.Receipt, error) {
+
+	// First we need to apply execution and postop messages to a copy, to be able to revert if
+	statedbCopy := statedb.Copy()
+	executionResult, paymasterPostOpResult, err := applyRip7560ExecutionAndPostOpMessages(config, vpr, bc, author, gp, statedbCopy, header, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// PostOp failed, reverting execution changes
+	if paymasterPostOpResult != nil && paymasterPostOpResult.Err != nil {
+		log.Error("postPaymasterTransaction reverted, not applying execution and postop")
+	} else {
+		// execution & postOp succeeded, applying to original stateDb
+		executionResult, paymasterPostOpResult, err = applyRip7560ExecutionAndPostOpMessages(config, vpr, bc, author, gp, statedb, header, cfg)
```

Comment:
not sure why you need to apply twice. nobody pays the node to run execution twice.